### PR TITLE
Ensure IMC sessions are logged out at each exit point (#65169)

### LINKED ISSUE #65169: Ansible 2.9.1: "Maximum sessions reached" error with imc_rest module
##### SUMMARY
After an upgrade from 2.8.4 to 2.9.1, I noticed a regression issue with the imc_rest module.
The module doesn't clean up its session after each connection to the UCS so after few tasks it won't be able to connect because there are too many session still opened on the CIMC server.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

imc_rest
##### ANSIBLE VERSION

```paste below
ansible 2.9.1
  config file = /root/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  7 2019, 00:51:29) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

##### CONFIGURATION

```paste below
DEFAULT_TIMEOUT(/root/ansible/ansible.cfg) = 30
PARAMIKO_HOST_KEY_AUTO_ADD(/root/ansible/ansible.cfg) = True
```

##### OS / ENVIRONMENT

The target host is a UCS-E server
CIMC version is 3.2(6.20180817145819)

##### STEPS TO REPRODUCE

Execute at least 5 tasks on the CIMC server

```yaml
- name: Set imc connection settings
  set_fact:
    hostname: "{{ hostvars[kickstart_iso_host].imc_ip }}"
    username: "{{ imc_username }}"
    password: "{{ hostvars[kickstart_iso_host].imc_password }}"

- name: Initialize Virtual Drive
  imc_rest:
    hostname: "{{ hostname }}"
    username: "{{ username }}"
    password: "{{ password }}"
    validate_certs: no
    content: |
        <configConfMo><inConfig>
           <storageVirtualDrive dn="sys/rack-unit-1/board/{{ ucs_virtual_drive_path }}" adminAction="start-fast-initialization"/>
        </inConfig></configConfMo>

- name: delete fanhe Vmedia mapping
  register: error
  ignore_errors: True
  changed_when:
    - '"No such volume" in error'
  imc_rest:
    hostname: "{{ hostname }}"
    username: "{{ username }}"
    password: "{{ password }}"
    validate_certs: no
    content: |
        <configConfMo><inConfig>
           <commVMediaMap dn="sys/svc-ext/vmedia-svc/vmmap-fanhe" volumeName="fanhe" map="www" status="deleted"/>
        </inConfig></configConfMo>

- name: delete all saved Vmedia mapping
  imc_rest:
    hostname: "{{ hostname }}"
    username: "{{ username }}"
    password: "{{ password }}"
    validate_certs: no
    content: |
        <configConfMo><inConfig>
           <commVMedia dn="sys/svc-ext/vmedia-svc" adminAction="delete-all-saved-mappings"/>
        </inConfig></configConfMo>

- name: Configure vMedia to point to ISO image on http server, ignore if already exists
  register: error
  ignore_errors: True
  changed_when:
    - '"already exist" in error'
  imc_rest:
    hostname: "{{ hostname }}"
    username: "{{ username }}"
    password: "{{ password }}"
    validate_certs: no
    content: |
        <configConfMo><inConfig>
           <commVMediaMap dn="sys/svc-ext/vmedia-svc/vmmap-fanhe" map="www"
            mountOptions="username=,password=" remoteFile="{{ kickstart_iso_host }}.iso"
            remoteShare="http://{{ workstation_ip }}/" status="created" volumeName="fanhe"/>
        </inConfig></configConfMo>

- name: Configuring Boot Order to  CDROM, HDD, FDD, PXE, and EFI Without Rebooting the Host
  imc_rest:
    hostname: "{{ hostname }}"
    username: "{{ username }}"
    password: "{{ password }}"
    validate_certs: no
    content: |
        <configConfMo><inConfig>
           <lsbootDef dn="sys/rack-unit-1/boot-policy" rebootOnUpdate="no">
           <lsbootVirtualMedia access="read-only" order="1" type="virtual-media" rn="vm-read-only"/>
           <lsbootVirtualMedia access="read-write" order="3" type="virtual-media" rn="vm-read-write"/>
           <lsbootLan rn="lan-read-only" access="read-only" order="4" prot="pxe" type="lan"/>
           <lsbootStorage rn="storage-read-write" access="read-write" order="2" type="storage"/>
           <lsbootEfi rn="efi-read-only" access="read-only" order="5" type="efi"/>
           </lsbootDef>
        </inConfig></configConfMo>
```

##### EXPECTED RESULTS

Any task should be executed properly and only one session should be opened at a time with the CIMC server.

Below are some on-going capture of the number of sessions opened on the CIMC server while the playbook is being executed. This is with 2.8.4 where it does work as expected:

```
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
29     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
29     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
30     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
30     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
30     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
31     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
21     admin            10.195.123.15     CLI          no
E1120D-FOC22242ZRX#
```

##### ACTUAL RESULTS

the 5th task will fail with "Maximum sessions reached" error
If you ssh to the CIMC server, you can see there are already 5 sessions active which is the limit of the server

```paste below
TASK [./roles/deploy-iso-image : Set imc connection settings] **************************************************************************************************************************************************************************
task path: /root/ansible/roles/deploy-iso-image/tasks/deploy-iso.yml:4
ok: [localhost] => {"ansible_facts": {"hostname": "10.195.123.28/24", "password": "password", "username": "admin"}, "changed": false}

TASK [./roles/deploy-iso-image : Initialize Virtual Drive] *****************************************************************************************************************************************************************************
task path: /root/ansible/roles/deploy-iso-image/tasks/deploy-iso.yml:10
changed: [localhost] => {"aaaLogin": {"attributes": {"cookie": "", "outCookie": "1574355909/7b8b0746-97de-17de-8006-e44f02735d00", "outPriv": "admin", "outRefreshPeriod": "600", "outSessionId": "8", "outVersion": "3.2(6.20180817145819)", "response": "yes"}}, "changed": true, "configConfMo": {"attributes": {"cookie": "1574355909/7b8b0746-97de-17de-8006-e44f02735d00", "dn": "", "response": "yes"}, "children": [{"outConfig": {"attributes": {}, "children": [{"storageVirtualDrive": {"attributes": {"accessPolicy": "read-write", "adminAction": "no-op", "allowBackgroundInit": "true", "autoDeleteOldest": "true", "autoSnapshot": "false", "bootDrive": "true", "cachePolicy": "direct-io", "currentWriteCachePolicy": "write-through", "diskCachePolicy": "unchanged", "dn": "sys/rack-unit-1/board/storage-SAS-SLOT-5/vd-0", "driveState": "Optimal", "drivesPerSpan": "2", "fdeCapable": "no", "fdeEnabled": "no", "health": "Good", "hotspareAction": "no-op", "id": "0", "name": "RAID1_01", "physicalDrivesList": "Physical drives to be used for virtual drive reconstruction. pd_id1,pd_id2,...pd_idN", "raidLevel": "RAID 1", "readPolicy": "no-read-ahead", "requestedWriteCachePolicy": "write-through", "size": "3662109 MB", "spanDepth": "1", "status": "modified", "stripSize": "64k", "targetId": "0", "vdStatus": "Optimal"}}}]}}]}, "elapsed": 8, "response": "OK (unknown bytes)", "status": 200}

TASK [./roles/deploy-iso-image : delete fanhe Vmedia mapping] **************************************************************************************************************************************************************************
task path: /root/ansible/roles/deploy-iso-image/tasks/deploy-iso.yml:21
ok: [localhost] => {"aaaLogin": {"attributes": {"cookie": "", "outCookie": "1574355918/4784411f-97df-17df-8007-e44f02735d00", "outPriv": "admin", "outRefreshPeriod": "600", "outSessionId": "9", "outVersion": "3.2(6.20180817145819)", "response": "yes"}}, "changed": false, "configConfMo": {"attributes": {"cookie": "1574355918/4784411f-97df-17df-8007-e44f02735d00", "dn": "", "response": "yes"}, "children": [{"outConfig": {"attributes": {}}}]}, "elapsed": 7, "response": "OK (unknown bytes)", "status": 200}

TASK [./roles/deploy-iso-image : delete all saved Vmedia mapping] **********************************************************************************************************************************************************************
task path: /root/ansible/roles/deploy-iso-image/tasks/deploy-iso.yml:36
changed: [localhost] => {"aaaLogin": {"attributes": {"cookie": "", "outCookie": "1574355926/c4cf9784-97de-17de-8008-e44f02735d00", "outPriv": "admin", "outRefreshPeriod": "600", "outSessionId": "10", "outVersion": "3.2(6.20180817145819)", "response": "yes"}}, "changed": true, "configConfMo": {"attributes": {"cookie": "1574355926/c4cf9784-97de-17de-8008-e44f02735d00", "dn": "", "response": "yes"}, "children": [{"outConfig": {"attributes": {}, "children": [{"commVMedia": {"attributes": {"activeSessions": "0", "adminAction": "no-op", "adminState": "enabled", "dn": "sys/svc-ext/vmedia-svc", "encryptionState": "disabled", "lowPowerUsbState": "enabled", "status": "modified"}}}]}}]}, "elapsed": 7, "response": "OK (unknown bytes)", "status": 200}

TASK [./roles/deploy-iso-image : Configure vMedia to point to ISO image on http server, ignore if already exists] **********************************************************************************************************************
task path: /root/ansible/roles/deploy-iso-image/tasks/deploy-iso.yml:47
ok: [localhost] => {"aaaLogin": {"attributes": {"cookie": "", "outCookie": "1574355934/d78b7e1c-97de-17de-8009-e44f02735d00", "outPriv": "admin", "outRefreshPeriod": "600", "outSessionId": "11", "outVersion": "3.2(6.20180817145819)", "response": "yes"}}, "changed": false, "configConfMo": {"attributes": {"cookie": "1574355934/d78b7e1c-97de-17de-8009-e44f02735d00", "dn": "", "response": "yes"}, "children": [{"outConfig": {"attributes": {}, "children": [{"commVMediaMap": {"attributes": {"adminAction": "no-op", "dn": "sys/svc-ext/vmedia-svc/vmmap-fanhe", "map": "www", "mappingStatus": "OK", "mountOptions": "noauto,username=,********=", "password": "", "remoteFile": "fanhe-ucse-2.iso", "remoteShare": "http://10.195.123.15/", "status": "created", "volumeName": "fanhe"}}}]}}]}, "elapsed": 8, "response": "OK (unknown bytes)", "status": 200}

TASK [./roles/deploy-iso-image : Configuring Boot Order to  CDROM, HDD, FDD, PXE, and EFI Without Rebooting the Host] ******************************************************************************************************************
task path: /root/ansible/roles/deploy-iso-image/tasks/deploy-iso.yml:64
fatal: [localhost]: FAILED! => {"aaaLogin": {"attributes": {"cookie": "", "errorCode": "554", "errorDescr": "Maximum sessions reached", "invocationResult": "unidentified-fail", "response": "yes"}}, "changed": false, "error_code": "554", "error_text": "Maximum sessions reached", "msg": "Request failed: Maximum sessions reached", "output": "<aaaLogin cookie=\"\" response=\"yes\" errorCode=\"554\" invocationResult=\"unidentified-fail\" errorDescr=\"Maximum sessions reached\"> </aaaLogin>\n"}
```

Below is the list of sessions opened on the CIMC server after the playbook failed:
```
E1120D-FOC22242ZRX# show user-session
ID     Name             IP Address        Type         Killable
------ ---------------- ----------------- ------------ --------
25     admin            10.195.123.15     xmlapi       yes
24     admin            10.195.123.15     xmlapi       yes
23     admin            10.195.123.15     xmlapi       yes
22     admin            10.195.123.15     xmlapi       yes
21     admin            10.195.123.15     CLI          no
19     admin            10.61.204.131     webgui       yes
```

### PR DESCRIPTION
**Pull request by etrikp**
##### SUMMARY
Fixes #65169

The atexit hook alone is not sufficient enough to ensure all IMC sessions are close properly. The atexit module is hooked to the exit of the python interpreter and the imc_rest module itself can exit before the interpreter exits.  To fix this, calls to the logout function were added to each point where an exit happens and an IMC session has been established.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
imc_rest

### CODE REVIEW

**Comment by dagwieers** [comment]
In the past this wasn't needed, so why did this behaviour change?
It means Ansible modules no longer work with atexit? That doesn't seem right.

**Comment by etrikp** [comment]
I have always experienced this issue with imc_rest, all the way back to 2.6. It tends to happen when iterating through failures with plays calling imc_rest multiple times. Considering these lines 

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/remote_management/imc/imc_rest.py#L412
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/remote_management/imc/imc_rest.py#L427

If I have a play that has more than 4 calls to imc_rest (XMLapi max session has a hard limit at 4 in IMC) before the python interpreter exits, the session limit will be reached and the API will refuse new connections. Its not so much that atexit is not working or has a bug, its just that the session limit is reached because of how many times imc_rest is called.

Talking through this, I think there is an even better way to deal with it and there is an optimization I can put in place so that only one session is needed per play regardless of how many calls to imc_rest are made. Thanks for questioning the changes! I will work that change up and resubmit the PR.

**Comment by dagwieers** [comment]
If it does not logout when python exits, then `atexit` is not working. That is the whole point of using `atexit` here, things may fail unexpectedly and we should `logout()` at every exit scenario. So your fix is not fixing all the uses of a working `atexit` and we should be looking at the cause.

File: lib/ansible/modules/remote_management/imc/imc_rest.py [hunk 1]
```diff
@@ -408,6 +408,7 @@ def main():
         # Perform actual request
         resp, info = fetch_url(module, url, data=data, method='POST', timeout=timeout)
         if resp is None or info['status'] != 200:
+            logout(module, url, cookie, timeout)
```
Comment (dagwieers) [inline lib/ansible/modules/remote_management/imc/imc_rest.py:411]: This shouldn't be needed, as `atexit` takes care of it. In fact, `atexit` also takes care of unexpected failures by logging out properly.

**Review by dagwieers — DISMISSED** [review]

**Comment by etrikp** [comment]
Again, the issue is not with atexit, the issue is with multiple calls to imc_rest exhausting the available sessions for the XLMApi. If I call imc_rest 5 times in one play, it tries to create 5 sessions and fails to create the 5th.

**Comment by dagwieers** [comment]
> I have always experienced this issue with imc_rest, all the way back to 2.6. It tends to happen when iterating through failures with plays calling imc_rest multiple times.

It doesn't make sense, so you are saying that `atexit` does work when it exits successfully, but not when it fails. `atexit` is ignorant of the exit scenario AFAICT.

Are you by any chance running something like Mitogen?

**Comment by dagwieers** [comment]
> atexit is only called when the python process is finished, not when an ansible play or module execution is finished.

Oh, right. But then we have a problem when there is an unexpected exit. I guess what we need is a way to hook into the module exit. This would require changes to module_utils to allow this, as the module could not handle this itself.

**Comment by dagwieers** [comment]
BTW you can already do multiple calls with a single `imc_rest` call today.

**Comment by etrikp** [comment]
The edge case we are talking through is not something that would come up if you bundle multiple payloads into a single imc_rest call, so thats likely why its hit or miss for some.. In the case where I call imc_rest 5 times in one play, and there are no failures on the first 4 invocations of imc_rest, the 5th call will fail because the previous 4 calls did not clean up their sessions.

For my use case, the more transactional approach to calling imc_rest suits the needs better.

It would be nice of there were some helpers to hook to an exit of a module, but this is a bigger change, and I don't see why adding the logout call to points where the module finishes is bad. The atexit hook is more for cleaning up after unexpected failures, not cleanup after a successful run.

**Comment by dagwieers** [comment]
Well, if `atexit` would be working (like before when we did not reuse the python interpreter) it worked fine for all exit scenarios. With this PR there would be 2 calls to `logout()` in some scenarios.

**Comment by etrikp** [comment]
Yup spot on, if each module was invoked in its own interpreter, the issue would not exist. I'll look further into the scenario of multiple calls to logout as well.

**Comment by Akasurde** [comment]
@etrikp Could you please migrate this PR to [`community.general`](https://github.com/ansible-collections/community.general) since the module is migrated to this repo. Thanks.

needs_info

**Pull request closed without merging**